// @stencila/schema/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ComputerLanguage_1 = require("./ComputerLanguage");
exports.ComputerLanguage = ComputerLanguage_1.default;
var CreativeWork_1 = require("./CreativeWork");
exports.CreativeWork = CreativeWork_1.default;
var Intangible_1 = require("./Intangible");
exports.Intangible = Intangible_1.default;
var Organization_1 = require("./Organization");
exports.Organization = Organization_1.default;
var OperatingSystem_1 = require("./OperatingSystem");
exports.OperatingSystem = OperatingSystem_1.default;
var Person_1 = require("./Person");
exports.Person = Person_1.default;
var SoftwareApplication_1 = require("./SoftwareApplication");
exports.SoftwareApplication = SoftwareApplication_1.default;
var SoftwareEnvironment_1 = require("./SoftwareEnvironment");
exports.SoftwareEnvironment = SoftwareEnvironment_1.default;
var SoftwarePackage_1 = require("./SoftwarePackage");
exports.SoftwarePackage = SoftwarePackage_1.default;
var SoftwareSession_1 = require("./SoftwareSession");
exports.SoftwareSession = SoftwareSession_1.default;
var SoftwareSourceCode_1 = require("./SoftwareSourceCode");
exports.SoftwareSourceCode = SoftwareSourceCode_1.default;
var Thing_1 = require("./Thing");
exports.Thing = Thing_1.default;
